### ISSUE TITLE
CI: starting user pod has intermittency failures

### ISSUE THREAD

**Issue by consideRatio** [jupyterhub/zero-to-jupyterhub-k8s#1858]
## Call to hub api to start a user-server fail to start
This is a bug I have no clue what it is about, but have seen show up here and there the last weeks when the CI tests run. Perhaps it is related to kubespawner being bumped, or was it erroring like this before as well? I'm not sure.

#### A build failure
[Build failure with logs of hub and proxy pod](https://travis-ci.org/github/jupyterhub/zero-to-jupyterhub-k8s/jobs/737443531) but not for the autohttps pod.

#### Example user_model seen
```python
# example user_model that is seen during failures
user_model = {'kind': 'user', 'name': 'testuser-f8ff861a-538a-4c93-a864-ae3345cea907', 'admin': False, 'groups': [], 'server': None, 'pending': None, 'created': '2020-10-20T14:28:03.524748Z', 'last_activity': '2020-10-20T14:28:11.990951Z', 'servers': {}, 'auth_state': None}
```

#### Code that fail to reach return statement
https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/2f7135cfeed631553eb3452f6b3f47f27e47c9f9/tests/test_spawn.py#L217-L224

## Logs inspection
The recurring theme is that the hub errors when speaking with the CHP proxy after the user has got a pod to represent it.

```
[I 2020-10-20 16:48:25.540 JupyterHub base:893] User testuser-aa6aa5ea-64bb-4719-93c7-04557f25c3d5 took 2.179 seconds to start
[I 2020-10-20 16:48:25.540 JupyterHub proxy:257] Adding user testuser-aa6aa5ea-64bb-4719-93c7-04557f25c3d5 to proxy /user/testuser-aa6aa5ea-64bb-4719-93c7-04557f25c3d5/ => http://192.168.79.75:8888
[D 2020-10-20 16:48:25.540 JupyterHub proxy:765] Proxy: Fetching POST http://proxy-api:8001/api/routes/user/testuser-aa6aa5ea-64bb-4719-93c7-04557f25c3d5
[I 2020-10-20 16:48:25.541 JupyterHub log:181] 200 GET /hub/api (@10.20.0.120) 1.70ms
[E 2020-10-20 16:48:25.542 JupyterHub base:915] Failed to add testuser-aa6aa5ea-64bb-4719-93c7-04557f25c3d5 to proxy!
    Traceback (most recent call last):
      File "/usr/local/lib/python3.8/dist-packages/jupyterhub/handlers/base.py", line 908, in finish_user_spawn
        await self.proxy.add_user(user, server_name)
      File "/usr/local/lib/python3.8/dist-packages/jupyterhub/proxy.py", line 270, in add_user
        await self.add_route(
      File "/usr/local/lib/python3.8/dist-packages/jupyterhub/proxy.py", line 781, in add_route
        await self.api_request(path, method='POST', body=body)
      File "/usr/local/lib/python3.8/dist-packages/jupyterhub/proxy.py", line 773, in api_request
        result = await client.fetch(req)
    tornado.curl_httpclient.CurlError: HTTP 599: Send failure: Broken pipe
```

If everything were as it should, the proxy should add a network route towards the newly added pod - it is a configurable proxy that JupyterHub controls after all. This is the kind of entry we should see twice because we have two tests adding a pod.

```
16:48:18.515 [ConfigProxy] info: Adding route /user/testuser-c2521e10-e843-40c3-8cb6-dac422b1433a -> http://192.168.79.74:8888
```

It seems like this never happens, and that the proxy pod / CHP doesn't notice any incoming request to add a route, because it doesn't log any incoming POST request either.

**Comment by consideRatio** [comment]
## Update

- The autohttps pod is irrelevant to consider, what fails is the hub pod's request directly to the proxy pod's API without going through the autohttps pod.
- The broken pipe error typically happen when starting a pod as part of the netpol test, which is the second time a pod is started and gets it route added. I have also seen it occur on the first user pod though, so I think that we observe it mostly as part of the netpol test relates to a timing - that it happens a bit later.

My main suspicion is that the hub pod has some kind of connection which is timing out after a while, which at the time cause the broken pipe error. I have some ideas on what can be done about this.

1. We make the hub->proxy connection more robust somehow - I lack knowledge about how still but perhaps something about sending keep alive messages?
2. We make the hub->proxy add route operation more robustly - for example by attempting again on 599 failure.

In other words, we need to fix something in jupyterhub to avoid the issue surfacing in our CI tests of Z2JH I think.
